tortoise/rerun.go

```diff
@@ -0,0 +1,147 @@
+package tortoise
+
+import (
+	"context"
+	"fmt"
+	"time"
+	"unsafe"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+)
+
+type rerunResult struct {
+	Consensus *turtle
+	Tracer    *validityTracer
+}
+
+func (t *ThreadSafeVerifyingTortoise) updateFromRerun(ctx context.Context) (bool, types.LayerID) {
+	var (
+		logger   = t.logger.WithContext(ctx).With()
+		reverted bool
+		observed types.LayerID
+	)
+	if value := t.update.Load(); value != nil {
+		var (
+			completed = (*rerunResult)(value)
+			current   = t.trtl
+			updated   = completed.Consensus
+			err       error
+		)
+		if current.Last.After(updated.Last) {
+			start := time.Now()
+			logger.Info("tortoise received more layers while rerun was in progress. running a catchup",
+				log.FieldNamed("last", current.Last),
+				log.FieldNamed("rerun-last", updated.Last))
+
+			err = catchupToCurrent(ctx, current, updated)
```

Comment:
is this necessary? 
if rerun decide we want to revert, is catching up to current last layer change that decision? 
if rerun decide no revert, then we wait till next rerun to cover this range.
i think the logic is simpler just exit here.

Reply:
it is not about revert. i can't exit here. after rerun completes we need to have votes for the same number of layers as original tortoise, for example:
1. rerun started when last was 20
2. original received 21
3. rerun completed with 20
4. original replaced with rerun
5. original(rerun) receives 22

Reply:
as time goes on, rerun is going to take longer and longer. this means that lag btwn original and rerun will be larger. overtime the original rerun problem will show up again (holding the lock for too long).

when catch-up fails, should we revert to the failed layer?

one thing i didn't understand about the original design was: why do we have to swap out the current tortoise when there is no revert? if we find nothing is wrong with the current tortoise, why not keep using it?

Reply:
> as time goes on, rerun is going to take longer and longer. this means that lag btwn original and rerun will be larger. overtime the original rerun problem will show up again (holding the lock for too long).

this is true, but not a problem. by my estimate rerun after 1 year will take around 30-40 minutes, assuming that we will have layer per minute that's only 40 layers. it will take less than second to do a catchup for 40 layers. even if it will be 60 layers after a year it is still less then a second. 600 layers after 10 years - will take around 5 seconds. this is assuming that we won't optimize tortoise any further. but i am pretty sure that once unified blocks lands rerun will be blazingly fast.

> when catch-up fails, should we revert to the failed layer?

i think we should continue using original state or terminate the application. i think it can only fail due to the bugs or system issues.

> one thing i didn't understand about the original design was: why do we have to swap out the current tortoise when there is no revert? if we find nothing is wrong with the current tortoise, why not keep using it?

i thought that swapping with the fresh state may hide some bugs that we have. but yes it shouldn't matter if we are confident, i'd rather leave it as it is for now though.

Reply:
> i thought that swapping with the fresh state may hide some bugs that we have. but yes it shouldn't matter if we are confident, i'd rather leave it as it is for now though.

it definitely hide potential bugs. and we will grow to rely on the rerun too much, which then may cause unnecessary turmoil in the network caused by regular reverts. IMO it's better to leave the state as is. but i agree it can be done later.

Reply:
> as time goes on, rerun is going to take longer and longer. this means that lag btwn original and rerun will be larger. overtime the original rerun problem will show up again (holding the lock for too long).

also once we have mesh state sync we probably won't need to do full reruns. if for example we know that aggregated hash is not equal starting from layer 100, we don't need to rerun anything for layers below that, right?

Reply:
> also once we have mesh state sync we probably won't need to do full reruns. if for example we know that aggregated hash is not equal starting from layer 100, we don't need to rerun anything for layers below that, right?

yes. agreed.